### DISCUSSION TITLE
Fixing Ulauncher after upgrading Python on Arch Linux

### DISCUSSION THREAD

**Discussion by friday** [Ulauncher/Ulauncher#1280] [Troubleshooting]
Arch Linux installs Python packages per Python version to `/usr/lib/python{PYTHON__VERSION/site-packages/{PACKAGE_NAME}`. These packages can normally only be imported for the specific Python versions they are installed.

When Arch upgrades Python versions (like 3.11 to 3.12), it also reinstalls all python system packages to the new directory. But it doesn't reinstall AUR packages, so you have to do that yourself.

To do that, you also need to make sure you do a _clean_ reinstall. One of these commands should do it, depending on if you use an AUR helper or not:
`paru -S ulauncher --rebuild`
`yay -S ulauncher --answerclean All`
`git clone https://aur.archlinux.org/ulauncher.git && cd ulauncher && makepkg -isC`

~~**Replace ulauncher with ulauncher-git** if you are using the ulauncher-git package (for the last command, replace in the git clone command, not the cd command)~~  ulauncher-git uses Ulauncher v6, which has a workaround for this issue by looking at older python site-packages.

To get a list of other packages affected after upgrading Python versions, you can run `pacman -Qoq /usr/lib/python{PREVIOUS_PYTHON_VERSION}`

References:
https://wiki.archlinux.org/title/python#Module_not_found_after_Python_version_update
https://www.reddit.com/r/archlinux/comments/rf6c84/psa_python_310_is_in_core_rebuild_your_aur/

**Comment by jjcortes22** [comment]
Thank you.

**Comment by vizagite** [comment]
I faced this recently bcoz I stopped updating AUR packages regularly as system packages.

Then I recalled I used to rebuild all such python conflict packages from updates with my paru as, say when I got python 3.13 in recent system update but I was using 3.12 earlier

`$ paru -S $(pacman -Qoq /usr/lib/python3.12) --rebuild`

basically same as friday's command for ulauncher but we grab all such AUR packages relying on older python and rebuild and install all of them in one go. just change python3.12 to pythonX based on your last version or seeing the `/usr/lib` folder

its about time, I should add this directly to pipeline of my non-AUR system update alias command now.

---
ok my new bashrc function done

```bash
parur() {
  # Step 1: Update system and get currently installed Python major.minor
  paru -Syu --repo || return
  local cur_ver
  cur_ver=$(python -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')

  # Step 2: Rebuild AUR packages for older Python versions
  for dir in /usr/lib/python*/; do
    local ver
    ver=$(basename "$dir" | sed 's/^python//; s/\.[0-9]*$//')  # major.minor only
    
    if [[ "$ver" != "$cur_ver" ]]; then
      # Get packages in the directory
      local packages
      packages=$(pacman -Qoq "$dir")
      
      if [[ -n "$packages" ]]; then
        # Rebuild the AUR packages for the old version
        paru -S --rebuild $packages || echo "Failed to rebuild packages for Python $ver"
      else
        echo "No packages to rebuild for Python $ver"
      fi
    fi
  done
}
```
